miniconf/src/packed.rs

```diff
@@ -0,0 +1,249 @@
+use crate::{IntoKeys, Keys};
+use core::{
+    num::NonZeroUsize,
+    ops::{Deref, DerefMut},
+};
+
+/// A bit-packed representation of `TreeKey` indices.
+///
+/// The value consists of a number of (from storage MSB to LSB):
+///
+/// * Zero or more groups of variable bit length, concatenated, each containing
+///   the index at the given `TreeKey` level. The deepest level is last.
+/// * A set bit to mark the end of the used bits.
+/// * Zero or more cleared bits corresponding to unused index space.
+///
+/// [`Packed::EMPTY`] has the marker at the MSB.
+/// During [`Packed::push_lsb()`] the values are inserted with their MSB
+/// where the marker was and the marker moves toward the storage LSB.
+/// During [`Packed::pop_msb()`] the values are removed with their MSB
+/// aligned with the storage MSB and marker moves toward the storage MSB.
+///
+/// The representation is MSB aligned to make key `Ord` more natural and stable.
+/// The `Packed` key `Ord` matches the ordering of node on a depth-first tree
+/// traversal. New nodes can be added to the tree without changing the implicit
+/// encoding as long no new bits need to be allocated. Under this condition
+/// the mapping between indices and `Packed` representation is stable.
+///
+/// "Small numbers" in LSB-aligned representation can be obtained through
+/// [`Packed::into_lsb()`]/[`Packed::from_lsb()`] but don't have the ordering
+/// and stability properties.
+///
+/// `Packed` can be used to uniquely identify
+/// nodes in a [`crate::TreeKey`] using only a very small amount of bits.
+/// For many realistic `TreeKey`s a `u16` or even a `u8` is sufficient
+/// to hold a `Packed` in LSB notation. Together with the `Postcard` trait
+/// and `postcard` `serde` format, this then gives access to any node in a nested
+/// heterogeneous `Tree` with just a `u16` or `u8` as compact key and `[u8]` as
+/// compact value.
+///
+/// ```
+/// # use miniconf::Packed;
+///
+/// let mut p = Packed::EMPTY;
+/// let mut p_lsb = 1; // marker
+/// for (bits, value) in [(2, 3), (1, 0), (0, 0), (3, 5)] {
+///     p.push_lsb(bits, value).unwrap();
+///     p_lsb <<= bits;
+///     p_lsb |= value;
+/// }
+/// assert_eq!(p_lsb, 0b1_11_0__101);
+/// assert_eq!(p, Packed::from_lsb(p_lsb.try_into().unwrap()));
```

Comment:
Why does the user have to specify the marker when providing an LSB-aligned version? It seems like this could be implicitly understood to exist above the MSB of the bit representation of the data.

I think this is what was tripping me up previously - I expected the marker bit to be handled entirely internally.

Reply:
Aaahhh yeah I forgot we could have leading zeros in an LSB-aligned implementation. That makes sense, thanks.